Remove audio engine from exporting code

szeli1 opened the pull request:
Project folder needs export to its own audio and `AudioEngine` does not allow this.

This pull request removes `AudioEngine` and `AudioDevice` from `AudioFileDevice` and `ProjectRenderer`, so the exporting code does not depend on it. This way in the future a Project folder can provide its own audio buffers simplifying the exporting process.
`ProjectRenderer` was renamed to `ActiveRenderer` and now `RenderManager` handles the project specific code.

Needs testing.

Rossmaxx commented on include/ActiveRenderer.h:
```diff
@@ -61,11 +67,15 @@ class LMMS_EXPORT ProjectRenderer : public QThread
 
 
-	ProjectRenderer( const AudioEngine::qualitySettings & _qs,
+	ActiveRenderer( const AudioEngine::qualitySettings & _qs,
```
```suggestion
	ActiveRenderer(const AudioEngine::qualitySettings& _qs,
```

Style.

Rossmaxx commented on include/ActiveRenderer.h:
```diff
@@ -61,11 +67,15 @@ class LMMS_EXPORT ProjectRenderer : public QThread
 				ExportFileFormat _file_format,
-				const QString & _out_file );
-	~ProjectRenderer() override = default;
+				const QString & _out_file,
```
```suggestion
				const QString& _out_file,
```

Rossmaxx commented on include/AudioDevice.h:
```diff
@@ -90,26 +90,28 @@ class AudioDevice
 	// clear given signed-int-16-buffer
-	void clearS16Buffer( int_sample_t * _outbuf,
-							const fpp_t _frames );
+	static void clearS16Buffer( int_sample_t * _outbuf,
```
```suggestion
	static void clearS16Buffer(int_sample_t* _outbuf,
```

Rossmaxx commented on include/AudioFileDevice.h:
```diff
@@ -28,19 +28,23 @@
-			const ch_cnt_t _channels, const QString & _file,
-			AudioEngine* audioEngine );
-	~AudioFileDevice() override;
+			const QString & _file, const ch_cnt_t _channels,
```
```suggestion
			const QString& _file, const ch_cnt_t _channels,
```

Rossmaxx commented on include/AudioFileDevice.h:
```diff
@@ -66,10 +83,13 @@ class AudioFileDevice : public AudioDevice
 
 using AudioFileDeviceInstantiaton
-	= AudioFileDevice* (*)(const QString&, const OutputSettings&, const ch_cnt_t, AudioEngine*, bool&);
+	= AudioFileDevice* (*)(const OutputSettings&m, bool&, const QString&, const ch_cnt_t, const fpp_t);
```
&m ?

Rossmaxx commented on include/AudioDevice.h:
```diff
@@ -90,26 +90,28 @@ class AudioDevice
+	static int convertToS16(const SampleFrame* _ab,
 						const fpp_t _frames,
 						int_sample_t * _output_buffer,
+						ch_cnt_t channels,
```
I believe @michaelgregorius simplified out this channel count in #7156

michaelgregorius replied:
Yes, `SampleFrame` is explicitly stereo because tons of code has already assumed this before my changes in https://github.com/LMMS/lmms/pull/7156 and multi-channel (it was called surround sound in the old code) was not really supported anyway. So this interface might be a bit contradictory if the channel count is intended to describe the number of mono channels in the buffer.

On a side note (and not really related to this PR): Does this conversion really belong in the `AudioDevice` class? The functionality is not really related solely to audio devices and other code might also be interested in such conversions. It could be moved out with regards to separation of concerns.

szeli1 replied:
So I should I replace `channels` with "2"?

michaelgregorius replied:
Why have a parameter if it must always be 2 anyway? You can just remove the parameter and assume stereo channels in the method's implementation.

Rossmaxx commented on src/core/ActiveRenderer.cpp:
```diff
@@ -39,28 +39,28 @@ namespace lmms
 
-const std::array<ProjectRenderer::FileEncodeDevice, 5> ProjectRenderer::fileEncodeDevices
+const std::array<ActiveRenderer::FileEncodeDevice, 5> ActiveRenderer::fileEncodeDevices
 {
```
Look at the codefactor check below this line.

Rossmaxx commented on src/core/ActiveRenderer.cpp:
```diff
@@ -71,22 +71,29 @@ const std::array<ProjectRenderer::FileEncodeDevice, 5> ProjectRenderer::fileEnco
 	// Maybe one day the user can add own encoders inside the program.
 
-	FileEncodeDevice{ ProjectRenderer::ExportFileFormat::Count, nullptr, nullptr, nullptr }
+	FileEncodeDevice{ActiveRenderer::ExportFileFormat::Count, nullptr, nullptr, nullptr}
```
Codefactor here too.

Rossmaxx commented on src/core/ActiveRenderer.cpp:
```diff
@@ -157,7 +181,7 @@ void ProjectRenderer::startProcessing()
-void ProjectRenderer::run()
+void ActiveRenderer::run()
 {
 #if 0
```
I believe this code block is safe to remove.

Rossmaxx commented on src/core/ActiveRenderer.cpp:
```diff
@@ -172,31 +196,44 @@ void ProjectRenderer::run()
-
-	// Now start processing
-	Engine::audioEngine()->startProcessing(false);
+	qDebug("start running");
```
std::cerr perhaps? Or what's the best way to output? @JohannesLorenz

szeli1 replied:
`qDebug` will be removed

Rossmaxx replied:
Okk then

Rossmaxx commented on src/core/RenderManager.cpp:
```diff
@@ -204,5 +246,50 @@ void RenderManager::updateConsoleProgress()
 
-} // namespace lmms
\ No newline at end of file
+} // namespace lmms
```
Not really needed but it would be nice to add a blank line at the end of file.

Rossmaxx commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -130,17 +130,18 @@ void AudioDevice::renamePort( AudioPort * )
 		for( fpp_t frame = 0; frame < _frames; ++frame )
 		{
-			for( ch_cnt_t chnl = 0; chnl < channels(); ++chnl )
+			for( ch_cnt_t chnl = 0; chnl < channels; ++chnl )
```
```suggestion
			for(ch_cnt_t chnl = 0; chnl < channels; ++chnl)
```

Do check at similar cases in this file

Rossmaxx reviewed (commented):
Style review. Do look beyond what i posted here, and codefactor, add blank lines where there's a warning.

sakertooth commented:
> This pull request removes AudioEngine and AudioDevice from AudioFileDevice and ProjectRenderer, so the exporting code does not depend on it. This way in the future a Project folder can provide its own audio buffers simplifying the exporting process.
ProjectRenderer was renamed to ActiveRenderer and now RenderManager handles the project specific code.

I do not follow. What do you mean by a project folder providing its own audio buffers? And why the name change? What else are we going to export?

szeli1 commented:
> I do not follow. What do you mean by a project folder providing its own audio buffers? And why the name change? What else are we going to export?

#5990 currently exports audio into the project file. This is not ideal so in the future a project folder feature needs to be added. This means a `SampleClip`'s `SampleBuffer` needs to be exported into a file in the project folder. Currently the exporting process depends on `AudioEngine`. I have not found a way to export audio that is not from the `AudioEngine`. This pull request makes it so anything can export any audio easily. I changed the name because I removed the project specific code from `ProjectFolder`.

sakertooth commented:
> #5990 currently exports audio into the project file. This is not ideal so in the future a project folder feature needs to be added. This means a `SampleClip`'s `SampleBuffer` needs to be exported into a file in the project folder. Currently the exporting process depends on `AudioEngine`. I have not found a way to export audio that is not from the `AudioEngine`. This pull request makes it so anything can export any audio easily. I changed the name because I removed the project specific code from `ProjectFolder`.

I get that, but if we need to export samples *inside the project*, then `ProjectRenderer` is a fitting name. `ActiveRenderer` personally is not clear to me. I always understood `ProjectRenderer` as something that renders the project or something specific inside it. It doesn't have to render the whole project. It can render individual tracks as well. All we would need to do is add functionality that allows it to render specific sample clips or MIDI clips to a file. The name doesn't have to change at all.

What needs to change is how `AudioDevice` and `AudioFileDevice` work. The only thing that really needs to happen is that `AudioDevice` and `AudioFileDevice` should use abstract callback functions to retrieve the data they need, rather than always getting the data from `AudioEngine`, which is pretty much the idea you have here I think? You can pass these callback functions in their constructor. I would keep `ProjectRenderer` alone for the most part, unless it has coupling with the `AudioEngine` that isn't necessary and should be removed as well.

I don't have much time to go through a more thorough explanation of what I'm envisioning the code should look like, but hopefully the reviews can fix that.

sakertooth commented:
Also, to make review easier, maybe we should split this PR up with more specific goals.

szeli1 commented:
> `ProjectRenderer` is a fitting name

Okay, I will keep the original name.

> What needs to change is how AudioDevice and AudioFileDevice work. The only thing that really needs to happen is that AudioDevice and AudioFileDevice should use abstract callback functions to retrieve the data they need, rather than always getting the data from AudioEngine, which is pretty much the idea you have here I think? You can pass these callback functions in their constructor. I would keep ProjectRenderer alone for the most part, unless it has coupling with the AudioEngine that isn't necessary and should be removed as well.

I would rather avoid changing `AudioDevice`. `AudioDevice` depends on `AudioEngine` for a reason and lots of files would need to be updated unnecessarily (like `AudioOss` and `AudioPulseAudio`). That is why I removed `AudioDevice` from `AudioFileDevice`.
I do not understand what you mean by "abstract callback functions". In this PR `ProjectRenderer` uses a function pointer to get a buffer from somewhere. Placing this in `AudioFileDevice` would make every constructor more complicated. Instead `ProjectRenderer` passes the buffer directly to `AudioFileDevice` without function pointers in constructors.
`ProjectRenderer` made calls to `Song` functions and `AudioEngine` functions before this. This needs to be removed so exporting a `SampleBuffer` in the future will not call those functions. Leaving them in would result in complex code handling multiple exporting modes (export project and export sample instead of export buffer).

sakertooth commented:
Like I said, I don't have much time to go through a more thorough explanation of what I am envisioning. I simply don't want a situation where the design of the code is just as bad and is something we would still have to clean up.

> That is why I removed AudioDevice from AudioFileDevice.

I am not in complete disagreement with separating these two since the inheritance relationship was always something that was causing complexity. If the `AudioDevice` class was designed better, we wouldn't even have to do this really, but unfortunately that is not the case.

> I do not understand what you mean by "abstract callback functions". In this PR ProjectRenderer uses a function pointer to get a buffer from somewhere. Placing this in AudioFileDevice would make every constructor more complicated. Instead ProjectRenderer passes the buffer directly to AudioFileDevice without function pointers in constructors.

The sentences that followed from the first sentence here is what I meant, but for those function pointers to be in the constructor for `AudioFileDevice`. My idea was that an `AudioFileDevice` and `AudioDevice` do not need to know anything about the producer of the data it is consuming. It work independently from the producer. This makes the design a lot more flexible. The fact that there still might be coupling involved with `AudioEngine` and `AudioDevice` beyond just pulling the data from the buffer (e.g., `AudioDevice` needs to delete the buffer if it was returned from the FIFO thread) is a result a bad design and should be changed.

I am not saying that this is the design we should go with, I'm just sharing my thoughts and clarifying what I meant to say. It will take me a lot more time looking at the code and understanding what is necessary to come up with a more suitable design if one exists.

Also, please explain how function pointers inside a constructor make the constructor more complicated? The length of the function pointer isn't an issue since you can use an alias, and you really only need one callback function to achieve this.

> ProjectRenderer made calls to Song functions and AudioEngine functions before this. This needs to be removed so exporting a SampleBuffer in the future will not call those functions. Leaving them in would result in complex code handling multiple exporting modes (export project and export sample instead of export buffer).

It's not just about what functions it calls. Removing code just like that because of coupling and not understanding why it was there and being called is risky.

The fact that the exporting code might handle multiple exporting modes at once isn't necessarily a bad thing. Whether we are exporting the project, exporting a single sample or MIDI clip, or exporting a track or the tracks individually, it is fine for it to handle all of that at once as long as the design is clear and concise. It will only be complex if the design fails at a fundamental level and is not flexible to work in other situations, which is what we are dealing with in terms of the current code.

szeli1 commented:
> The sentences that followed from the first sentence here is what I meant, but for those function pointers to be in the constructor for AudioFileDevice. My idea was that an AudioFileDevice and AudioDevice do not need to know anything about the producer of the data it is consuming. It work independently from the producer.

My initial idea was this. Change `AudioDevice` so it doesn't depend on `AudioEngine`. The problem was that `AudioEngine` needed to be removed from at least 11 classes and I didn't know what half of them did (like `AudioOss` and `AudioPulseAudio`). It was too much work and I couldn't see a reason why `AudioEngine` needed to be removed from everywhere.

```
The sentences that followed from the first sentence here is what I meant, but for those function pointers to be in the constructor for AudioFileDevice. My idea was that an AudioFileDevice and AudioDevice do not need to know anything about the producer of the data it is consuming. It work independently from the producer. This makes the design a lot more flexible. The fact that there still might be coupling involved with AudioEngine and AudioDevice beyond just pulling the data from the buffer (e.g., AudioDevice needs to delete the buffer if it was returned from the FIFO thread) is a result a bad design and should be changed.

I am not saying that this is the design we should go with, I'm just sharing my thoughts and clarifying what I meant to say. It will take me a lot more time looking at the code and understanding what is necessary to come up with a more suitable design if one exists.
```
This was the approach I followed while implementing. I'm not pressuring you to look at the code and your points could be true.

> Also, please explain how function pointers inside a constructor make the constructor more complicated? The length of the function pointer isn't an issue since you can use an alias, and you really only need one callback function to achieve this.

It can be confusing and it isn't needed. Function pointers add an other layer of complexity. `AudioFileDevice`'s job is to write a buffer into a file. This can be done by a simple function such as `AudioFileDevice::processThisBuffer(SampleFrame* frameBuffer, const fpp_t frameCount))`, storing buffers and function pointers isn't needed. Passing in function pointers into 5 constructors doesn't need to happen. So in this PR `ProjectRenderer` deals with getting the buffer and `AudioFileDevice` writes that buffer. This means that only 1 class has function pointers instead of 6 classes. Why should 6 classes have function pointers in their constructors?

> It's not just about what functions it calls. Removing code just like that because of coupling and not understanding why it was there and being called is risky.

It wasn't removed, only moved. It made `RenderManager` slightly more complicated but this way `ProjectRenderer` is as clean as possible. `ProjectRenderer` job is to get the right encoding and format and to be able to export audio on a separate thread in my opinion. `RenderManager` already deals with project code and track code and this way it still can run the processing on a separated thread.

> The fact that the exporting code might handle multiple exporting modes at once isn't necessarily a bad thing. Whether we are exporting the project, exporting a single sample or MIDI clip, or exporting a track or the tracks individually, it is fine for it to handle all of that at once as long as the design is clear and concise. It will only be complex if the design fails at a fundamental level and is not flexible to work in other situations, which is what we are dealing with in terms of the current code.

In my opinion this pull request isn't complex and fixes most exporting problems. This can be true to your idea as well. I think this way there is more separation between `ProjectRenderer` and `RenderManager`. ProjectRenderer handles audio export while `RenderManager` handles project code.

I can see your Ideas but I have already tried to implement some of them (AudioDevice remove AudioEngine, AudioFileDevice function pointers) and realized that this is the cleanest way.
I think your best argument is whit `ProjectRenderer` handling project code, but I'm not convinced.
And thank you for your contribution to this discussion, I know it takes a lot of time.

sakertooth commented:
> In my opinion this pull request isn't complex and fixes most exporting problems.

> potentially closes https://github.com/LMMS/lmms/pull/7449

It probably doesn't since most of the exporting problems I believe involve starting and stopping multiple threads like the FIFO thread and the audio device thread and something goes wrong with the sequencing of these events (among other things), causing one thread to wait indefinitely for another to close. This is what I have observed looking through the problem in a debugger. AFAICT, this PR tries to remove coupling to allow for a more flexible design so we can export specific things inside the project, like a sample.

> It wasn't removed, only moved. It made RenderManager slightly more complicated but this way ProjectRenderer is as clean as possible. ProjectRenderer's job is to get the right encoding and format and to be able to export audio on a separate thread in my opinion. RenderManager already deals with project code and track code and this way it still can run the processing on a separated thread.

> I think this way there is more separation between ProjectRenderer and RenderManager. ProjectRenderer handles audio export while RenderManager handles project code.

`RenderManager`'s purpose wasn't to facilitate doing the actual rendering. It was to invoke the `ProjectRenderer` (which drives the actual rendering using the buffers directly) at the correct instances to either export the whole project, or just individual tracks. I have a problem with how these roles are suddenly being changed, especially if the design isn't being changed holistically. If it is, then working through the new design proposal first and sharing it is the way to go before any code is written.

Looking at the PR, it seems like you are only working around the current design very closely to allow for the new feature, which we know isn't working in terms of being scalable and flexible. If the design needs to change, so be it, but working through that new design is the first step to see if its an improvement in terms of being malleable enough to allow for other modifications we may want to apply to it in the future with ease. Otherwise, its just another thing we would have to revisit later.

szeli1 commented:
> It probably doesn't since most of the exporting problems I believe involve starting and stopping multiple threads like the FIFO thread and the audio device thread and something goes wrong with the sequencing of these events (among other things), causing one thread to wait indefinitely for another to close. This is what I have observed looking through the problem in a debugger.

I believe there is only one thread running. The `ProjectRenderer` thread. FIFO is stopped before exporting starts. When `ProjectRenderer` doesn't get a buffer it closes, so I think only the FIFO thread could stop it somehow. I can implement your PR, so this doesn't cause issues.

> RenderManager's purpose wasn't to facilitate doing the actual rendering.

> Looking at the PR, it seems like you are only working around the current design very closely to allow for the new feature, which we know isn't working in terms of being scalable and flexible.

So on it own in this PR `ProjectRenderer` doesn't depend on anything that would limit or control audio exporting. It doesn't run `Song::StartExport` for example. This means that on its own it does what it needs to do (In my opinion), it exports audio.
1.
It seems to me that you want to move back `Song` and `AudioEngine` in `ProjectRenderer` (from `RenderManager`) while you seem to see a bigger problem in the audio code design too. Why introduce large scale design problems by including `Song` in Project renderer instead of separating the two classes? Why allow the old bad design to effect core parts of code instead of removing the dependence on bad code?

> we know isn't working in terms of being scalable and flexible.

Yes, I agree with you about the current audio code isn't the best. This PR removes one part of the problem. `ProjectRenderer` depends only on `SampleFrame` (no design problems) instead of `AudioEngine` (design problems).
What does flexible mean in this case?
2.
It seems to me that you want to fix the whole audio system but you only want to fix specific parts one by one:

> I have a problem with how these roles are suddenly being changed, especially if the design isn't being changed holistically.
> Also, to make review easier, maybe we should split this PR up with more specific goals.

This PR makes `RenderManager` slightly wore and leaves `AudioEngine::doSetAudioDevice` without use. This isn't good. But I think it mostly fixes `ProjectRenderer`'s problems while being a smaller PR (only fixes a smaller part of the audio system). I believe this PR's benefits are bigger than its disadvantages (`RenderManager` thread running 6 lines of exporting code).

michaelgregorius commented:
> AFAICT, this PR tries to remove coupling to allow for a more flexible design so we can export specific things inside the project, like a sample.

@szeli1, @sakertooth: If the intention of this PR is indeed to enable something like the export of samples into files then the `ProjectRenderer` is IMO indeed the wrong place to do so. As it name implies it is used to render a whole project into a file and therefore it knows how to manage all the other involved entities to do so.

If something like a sample should be exported then there should be some simple to use class which you instantiate with a file name, then feed it buffers until done and the finish up the file when it is signaled that no more data will come. Interestingly, I have written just such a class to write raw audio files to debug another problem in LMMS. Perhaps I should create a PR for it as it is quite useful and could also serve as an example of the interface that I have described above.

This functionality should not be shoe-horned into `ProjectRenderer` just because it already knows how to write into a file.

Perhaps all the stuff related to `audioEncoderFactory` in `ProjectRenderer` could be extracted into a service class which then lets you provide a file name and an export format. Then you feed that things buffers until done. This class might then be shared between `ProjectRenderer` and the other code that wants to export samples, etc.

IMO this pull request has a problem with separations of concern.

sakertooth commented:
> Interestingly, I have written just such a class to write raw audio files to debug another problem in LMMS.

The problem I was thinking about with this solution is how you would handle rendering with insert effects/master effects. FL handles this quite well, and has options for both when rendering audio clips.

![image](https://github.com/user-attachments/assets/10180d29-8d61-4af4-aafb-c3cbfc2e1148)

The current audio infrastructure doesn't allow for this functionality to be implemented without difficulty I believe.

szeli1 commented:
> Perhaps all the stuff related to audioEncoderFactory in ProjectRenderer could be extracted into a service class which then lets you provide a file name and an export format. Then you feed that things buffers until done. This class might then be shared between ProjectRenderer and the other code that wants to export samples, etc.

This is the best solution I've heard yet. I'm willing to implement this.

> This functionality should not be shoe-horned into ProjectRenderer just because it already knows how to write into a file.

This sentence makes it sound way too bad to me. I have only moved about 10 lines of exporting code from `ProjectRenderer` to `RenderManager` and I moved a function.

> 
I have written just such a class to write raw audio files to debug another problem in LMMS. Perhaps I should create a PR for it as it is quite useful and could also serve as an example of the interface that I have described above.

So what should be done next? Close this PR and merge yours? I'm willing to implement those changes that you have requested.

sakertooth commented:
> > Interestingly, I have written just such a class to write raw audio files to debug another problem in LMMS.
> 
> 
> 
> The problem I was thinking about with this solution is how you would handle rendering with insert effects/master effects. FL handles this quite well, and has options for both when rendering audio clips.
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/10180d29-8d61-4af4-aafb-c3cbfc2e1148)
> 
> 
> 
> The current audio infrastructure doesn't allow for this functionality to be implemented without difficulty I believe.

Then again, I might be getting ahead of myself, so we might not need to worry about this now. There's a lot to change in the code and it's not feasible to do it all at once.

michaelgregorius commented:
> > Perhaps all the stuff related to audioEncoderFactory in ProjectRenderer could be extracted into a service class which then lets you provide a file name and an export format. Then you feed that things buffers until done. This class might then be shared between ProjectRenderer and the other code that wants to export samples, etc.
> 
> This is the best solution I've heard yet. I'm willing to implement this.

Thanks!

> > This functionality should not be shoe-horned into ProjectRenderer just because it already knows how to write into a file.
> 
> This sentence makes it sound way too bad to me. I have only moved about 10 lines of exporting code from `ProjectRenderer` to `RenderManager` and I moved a function.

Sorry if it sounded too harsh. Perhaps I have also misunderstood something but to me it sounded like many o the file exports should be routed through `ProjectRenderer`, e.g. also the ones that write samples to files, and that would be bad in my opinion. I got this impression because with this PR `ProjectRenderer` now takes a generic functor which provides the sample buffer and that basically means that it does not care anymore where the data comes from. But in fact it should "know" that the data comes from the render of the project because that's its task. So there shouldn't be any need for such a broad interface in `ProjectRenderer`.

> > I have written just such a class to write raw audio files to debug another problem in LMMS. Perhaps I should create a PR for it as it is quite useful and could also serve as an example of the interface that I have described above.
> 
> So what should be done next? Close this PR and merge yours? I'm willing to implement those changes that you have requested.

Please don't close this PR. The class that writes raw audio is rather small and I guess it does not really encompass what you had in mind with this PR or what @sakertooth had in mind.

michaelgregorius commented:
> > > Interestingly, I have written just such a class to write raw audio files to debug another problem in LMMS.
> > 
> > 
> > The problem I was thinking about with this solution is how you would handle rendering with insert effects/master effects. FL handles this quite well, and has options for both when rendering audio clips.
> > ![image](https://private-user-images.githubusercontent.com/20197645/358665363-10180d29-8d61-4af4-aafb-c3cbfc2e1148.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MzI0OTIsIm5iZiI6MTcyMzgzMjE5MiwicGF0aCI6Ii8yMDE5NzY0NS8zNTg2NjUzNjMtMTAxODBkMjktOGQ2MS00YWY0LWFhZmItYzNjYmZjMmUxMTQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDE4MTYzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhNzc4MzFhMzA1YjVmMjBlYzg0NDQ3NmFkYmRjNjg4NzdlNmM0NzRkZjJkOTU2MWZkYTBmYTY3MzkyZGUxOTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4W3tFZi20iOHGW6aMul9gVoceI_JM8sHFD6M7KXp0XM)
> > The current audio infrastructure doesn't allow for this functionality to be implemented without difficulty I believe.
> 
> Then again, I might be getting ahead of myself, so we might not need to worry about this now. There's a lot to change in the code and it's not feasible to do it all at once.

Bouncing audio should be something that should be designed very consciously. It should for example not be implemented like setting all other channels to mute, render the project and then undo everything. Instead it should be code that knows how to evaluate individual and isolated paths of the audio graph, render the path and write the result into a file.

I guess that FL studio does something very similar. In one case they stop the "graph traversal" before the master channel and in the other case the audio from the partial graph evaluation gets routed through the master effects as well.

There's lots of potential work to do with regards to the audio graph anyway. For example, if we want to support recording from different audio input sources then the source nodes of the graph, e.g. track inputs, will need to be connected to the different input buffers which have to be filled by the different audio drivers/devices. Due to this the `AudioEngine` likely cannot be removed from the audio devices anyway. Or the driver might be handed some buffer manager (which is likely also managed by the `AudioEngine` anyway) which will be filled during each processing callback/step.

sakertooth commented:
> Bouncing audio should be something that should be designed very consciously. It should for example not be implemented like setting all other channels to mute, render the project and then undo everything. Instead it should be code that knows how to evaluate individual and isolated paths of the audio graph, render the path and write the result into a file.

I agree. I recommended the former as an easy solution, knowing that having a proper audio graph implemented could take a while, even though exporting based on an audio graph would be an optimal solution.

sakertooth commented:
In that case, I would like having a proper audio graph first before we implement something like this. The solutions that don't use an audio graph could work, but can cause more headache down the line.

szeli1 commented:
> Sorry if it sounded too harsh.
Don't worry, it didn't offend me. You understood things correctly, I will move the`ProjectRenderer` code back in `ProjectRenderer`. I will implement an other class that is essentially `ProjectRenderer` currently (in this pr). It will run on it's own thread and it will handle exporting and getting/writing the buffers, just like `ProjectRenderer` does, with function pointers.

I'm glad that this seems to be resolved now.

michaelgregorius commented:
> In that case, I would like having a proper audio graph first before we implement something like this. The solutions that don't use an audio graph could work, but can cause more headache down the line.

By the way, is there any kind of documentation of how this currently works? There's got to be some code somewhere already that evaluates in which order the different tracks, channels, etc. can be evaluated so that everything works correctly, e.g. with regards to channel routing with sends.

sakertooth commented:
> By the way, is there any kind of documentation of how this currently works? There's got to be some code somewhere already that evaluates in which order the different tracks, channels, etc. can be evaluated so that everything works correctly, e.g. with regards to channel routing with sends.

The evaluation order currently works in stages. For the mixer stage, the code does have some kind of mechanism that resembles a graph relationship (by necessity), but everything else works in stages. The amount of documentation for this is somewhat sparse, but you can take a look [here](https://github.com/LMMS/lmms/wiki/LMMS-Architecture) for more info. The routing for mixer channels can be found in the `Mixer` class. Take note of the `MixerRoute` class and the dependency notification function at the end of `MixerChannel::doProcessing`.

Rossmaxx commented on include/AudioDevice.h:
```diff
@@ -90,26 +90,26 @@ class AudioDevice
 	// clear given signed-int-16-buffer
-	void clearS16Buffer( int_sample_t * _outbuf,
-							const fpp_t _frames );
+	static void clearS16Buffer(int_sample_t * _outbuf,
```
```suggestion
	static void clearS16Buffer(int_sample_t* _outbuf,
```

Rossmaxx commented on include/AudioFileOgg.h:
```diff
@@ -40,20 +40,24 @@ namespace lmms
+	static AudioFileDevice* getInst(
+			OutputSettings const &outputSettings,
+			bool &successful,
+			const QString & outputFilename,
```
```suggestion
			const QString& outputFilename,
```

Rossmaxx commented on include/AudioFileWave.h:
```diff
@@ -37,21 +37,24 @@ namespace lmms
+	static AudioFileDevice* getInst(
+			OutputSettings const &outputSettings,
+			bool &successful,
+			const QString & outputFilename,
```
```suggestion
			const QString& outputFilename,
```

Rossmaxx commented on src/core/AudioEngine.cpp:
```diff
@@ -240,6 +240,15 @@ void AudioEngine::stopProcessing()
 }
 
 
+void AudioEngine::startExporting(const struct qualitySettings& qs)
```
Shouldn't there be a finishExporting too?

Rossmaxx replied:
Else there's chances of lmms becoming unresponsive after export.

szeli1 replied:
Finish exporting would run `AudioEngine::stopProcessing()`. I could add in that function if you want.

Rossmaxx replied:
If on finish, processing is stopped, then why do you need to stop processing on starting export too?

Rossmaxx commented on src/core/ProjectRenderer.cpp:
```diff
@@ -242,5 +97,50 @@ void ProjectRenderer::updateConsoleProgress()
 	fflush( stderr );
 }
 
+// gets a buffer and some data as input
```
I think saker said on discord about this. The comment should go in the header file as 
```
/*
* First line
* Second line
* ... 
*/
``` 
format

szeli1 replied:
Shouldn't it look like this?:

```
/**
* First line
* Second line
* ... 
*/
```

And does this mean that I need to rewrite every comment I wrote in my PRs?

Rossmaxx replied:
Yeah this is what i meant, and regarding re writing, don't rewrite everything, just the ones you use to describe the function.

Rossmaxx commented on src/core/ProjectRenderer.cpp:
```diff
@@ -242,5 +97,50 @@ void ProjectRenderer::updateConsoleProgress()
+	if (newBuffer != nullptr && Engine::getSong()->isExportDone() == false)
+	{
+		// copy new buffer to bufferOut
+		for (size_t i = 0; i < bufferOut->size(); i++)
```
I believe there's a std::copy for this.

Rossmaxx commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -130,17 +130,19 @@ void AudioDevice::renamePort( AudioPort * )
+
+								const bool _convert_endian)
 {
+	const unsigned int channels = 2;
```
If this doesn't change at all, better make it constexpr

Rossmaxx commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -130,17 +130,19 @@ void AudioDevice::renamePort( AudioPort * )
 				auto temp = static_cast<int_sample_t>(AudioEngine::clip(_ab[frame][chnl]) * OUTPUT_SAMPLE_MULTIPLIER);
 
-				( _output_buffer + frame * channels() )[chnl] =
+				( _output_buffer + frame * channels )[chnl] =
```
```suggestion
				(_output_buffer + frame * channels)[chnl] =
```

Rossmaxx replied:
Btw what's this used for?

szeli1 replied:
I don't know.

But audio exporting classes use it.

Rossmaxx commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -150,26 +152,27 @@ int AudioDevice::convertToS16(const SampleFrame* _ab,
 
 
-void AudioDevice::clearS16Buffer( int_sample_t * _outbuf, const fpp_t _frames )
+void AudioDevice::clearS16Buffer(int_sample_t * _outbuf, const fpp_t _frames)
```
```suggestion
void AudioDevice::clearS16Buffer(int_sample_t* _outbuf, const fpp_t _frames)
```

Rossmaxx commented on src/core/audio/AudioFileWave.cpp:
```diff
@@ -95,25 +97,26 @@ bool AudioFileWave::startEncoding()
 	if( bitDepth == OutputSettings::BitDepth::Depth32Bit || bitDepth == OutputSettings::BitDepth::Depth24Bit )
 	{
-		auto buf = new float[_frames * channels()];
+		auto buf = new float[_frames * channels];
```
Std::array?

Rossmaxx reviewed (commented):
This is what I got for now.
